<launch>

  <arg name="ouster_ns" default="ouster" description="Override the default namespace of all ouster nodes"/>
  <arg name="timestamp_mode" default="" description="method used to timestamp measurements; possible values: {
    TIME_FROM_INTERNAL_OSC,
    TIME_FROM_SYNC_PULSE_IN,
    TIME_FROM_PTP_1588,
    TIME_FROM_ROS_TIME
    }"/>
  <arg name="metadata" description="path to write metadata file when receiving sensor data"/>
  <arg name="bag_file" description="file name to use for the recorded bag file"/>
  <arg name="viz" default="true" description="whether to run a rviz"/>
  <arg name="rviz_config" default="$(find-pkg-share ouster_ros)/config/viz.rviz" description="optional rviz config file"/>
  <arg name="tf_prefix" default="" description="namespace for tf transforms"/>

  <group>
    <!-- <push-ros-namespace namespace="$(var ouster_ns)"/> -->

    <node_container pkg="rclcpp_components" exec="component_container_mt" name="os_container" output="screen" namespace="">
      <composable_node pkg="ouster_ros" plugin="ouster_ros::OusterReplay" name="os_sensor">
        <param name="metadata" value="$(var metadata)"/>
      </composable_node>
      <composable_node pkg="ouster_ros" plugin="ouster_ros::OusterCloud" name="os_cloud">
        <param name="tf_prefix" value="$(var tf_prefix)"/>
        <param name="timestamp_mode" value="$(var timestamp_mode)"/>
      </composable_node>
      <composable_node pkg="ouster_ros" plugin="ouster_ros::OusterImage" name="os_image"/>
    </node_container>

    <node if="$(var viz)" pkg="rviz2" exec="rviz2" name="rviz2" output="screen"
      launch-prefix="bash -c 'sleep 5; $0 $@'" args="-d $(var rviz_config)"/>

  </group>

  <let name="_use_bag_file_name" value="$(eval '\'$(var bag_file)\' != \'b\'')"/>
  <!-- TODO: find out why this doesn't work when passing to record -->
  <!-- <let name="_topics_to_record" value="/imu_packets /lidar_packets"/> -->

  <executable if="$(var _use_bag_file_name)" output="screen"
    launch-prefix="bash -c 'sleep 3; $0 $@'"
    cmd="ros2 bag play $(var bag_file) --clock"/>

</launch>
